File: jpa-to-sql/src/main/java/de/muenchen/jpa/criteria/JpaCriteriaQuery.java
package de.muenchen.jpa.criteria;

import jakarta.persistence.criteria.*;
import jakarta.persistence.metamodel.EntityType;
import lombok.Getter;

import java.util.*;

@Getter
public class JpaCriteriaQuery<T> extends JpaAbstractQuery<T> implements CriteriaQuery<T> {

    private final List<Order> orderBys = new ArrayList<>();
    private Selection<T> selects = null;

    public JpaCriteriaQuery(Class<T> clazz, AbstractJpaExpressionFactory factory) {
        super(clazz, factory);
    }

    private Integer limit = null;
    private Integer offset = null;

    public CriteriaQuery<T> limit(Integer limit) {
        this.limit = limit;
        return this;
    }

    public CriteriaQuery<T> offset(Integer offset) {
        this.offset = offset;
        return this;
    }

    @Override
    public CriteriaQuery<T> select(Selection<? extends T> exp) {
        selects = (Selection<T>) exp;
        return this;
    }

    @Deprecated(since = "Deprecated in Java")
    @Override
    public CriteriaQuery<T> multiselect(Selection<?>... exp) {
        select(this.getFactory().construct(this.getClazz(), exp));
        return this;
    }

    @Deprecated(since = "Deprecated in Java")
    @Override
    public CriteriaQuery<T> multiselect(List<Selection<?>> selects) {
        if (selects != null) {
            multiselect(selects.toArray(new Selection[0]));
        }
        return this;
    }

    @Override
    public CriteriaQuery<T> orderBy(Order... exp) {
        orderBys.clear();
        orderBys.addAll(Arrays.stream(exp).filter(Objects::nonNull).toList());
        return this;
    }

    @Override
    public CriteriaQuery<T> orderBy(List<Order> exp) {
        if (exp != null) {
            return orderBy(exp.toArray(new Order[0]));
        }
        return this;
    }

    @Override
    public Selection<T> getSelection() {
        return selects;
    }

    @Override
    public List<Order> getOrderList() {
        return orderBys;
    }

    public <X> CriteriaQuery<T> from(CriteriaQuery<X> query) {
        throw new IllegalStateException("Not yet implemented");
    }

    @Override
    public CriteriaQuery<T> where(Expression<Boolean> exp) {
        return (CriteriaQuery<T>) super.where(exp);
    }

    @Override
    public CriteriaQuery<T> where(Predicate... exp) {
        return (CriteriaQuery<T>) super.where(exp);
    }

    @Override
    public CriteriaQuery<T> where(List<Predicate> exp) {
        return (CriteriaQuery<T>) super.where(exp);
    }

    @Override
    public CriteriaQuery<T> groupBy(Expression<?>... exp) {
        return (CriteriaQuery<T>) super.groupBy(exp);
    }

    @Override
    public CriteriaQuery<T> groupBy(List<Expression<?>> exp) {
        return (CriteriaQuery<T>) super.groupBy(exp);
    }

    @Override
    public CriteriaQuery<T> having(Expression<Boolean> exp) {
        return (CriteriaQuery<T>) super.having(exp);
    }

    @Override
    public CriteriaQuery<T> having(Predicate... exp) {
        return (CriteriaQuery<T>) super.having(exp);
    }

    @Override
    public CriteriaQuery<T> having(List<Predicate> exp) {
        return (CriteriaQuery<T>) super.having(exp);
    }

    @Override
    public CriteriaQuery<T> distinct(boolean d) {
        return (CriteriaQuery<T>) super.distinct(d);
    }
}

File: src/main/kotlin/de/muenchen/mostserver/data/jpa/JpaCriteriaQuery.kt
package de.muenchen.mostserver.data.jpa

import de.muenchen.mostserver.data.jpa.criteria.AbstractJpaExpressionFactory
import de.muenchen.mostserver.data.jpa.criteria.JpaPredicate
import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Expression
import jakarta.persistence.criteria.Order
import jakarta.persistence.criteria.ParameterExpression
import jakarta.persistence.criteria.Predicate
import jakarta.persistence.criteria.Root
import jakarta.persistence.criteria.Selection
import jakarta.persistence.criteria.Subquery
import jakarta.persistence.metamodel.EntityType

class JpaCriteriaQuery<T>(val clazz: Class<T>, val factory: AbstractJpaExpressionFactory): CriteriaQuery<T> {

    private var wheres: Predicate? = null
    private var selects: Selection<T?>? = null
    private val groupBys: MutableSet<Expression<*>> = HashSet()
    private var havings: Predicate? = null
    private val orderBys: MutableList<Order> = ArrayList()
    private val roots: MutableSet<Root<*>> = HashSet()
    private var isDistinct: Boolean = false

    var limit: Integer? = null;
    var offset: Integer? = null;

    fun limit(limit: Integer?): CriteriaQuery<T>? {
        this.limit = limit
        return this
    }

    fun offset(offset: Integer?): CriteriaQuery<T>? {
        this.offset = offset
        return this
    }

    override fun select(exp: Selection<out T?>?): CriteriaQuery<T>? {
        selects = exp as Selection<T?>
        return this
    }

    @Deprecated("Deprecated in Java")
    override fun multiselect(vararg exp: Selection<*>?): JpaCriteriaQuery<T>? {
        select(factory.construct(this.clazz, *exp.filterNotNull().toTypedArray()))
        return this
    }

    @Deprecated("Deprecated in Java")
    override fun multiselect(selects: List<Selection<*>?>?): CriteriaQuery<T>? {
        if (selects != null) {
            multiselect(*selects.toTypedArray())
        }
        return this
    }

    override fun where(exp: Expression<Boolean?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return where(JpaPredicate(factory, exp))
        }
        this.wheres = null
        return this
    }

    override fun where(vararg exp: Predicate?): CriteriaQuery<T> {
        if (exp.isEmpty()) {
            this.wheres = null;
        }
        this.wheres = factory.and(exp.filterNotNull())
        return this
    }

    override fun where(exp: List<Predicate?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return where(*exp.toTypedArray())
        }
        this.wheres = null
        return this
    }

    override fun groupBy(vararg exp: Expression<*>?): CriteriaQuery<T>? {
        exp.filterNotNull()
            .forEach(groupBys::add)
        return this
    }

    override fun groupBy(exp: List<Expression<*>?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return groupBy(*exp.toTypedArray())
        }
        return this
    }

    override fun having(exp: Expression<Boolean?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return having(JpaPredicate(factory, exp))
        }
        this.havings = null;
        return this
    }

    override fun having(vararg exp: Predicate?): CriteriaQuery<T>? {
        if (exp.isEmpty()) {
            this.havings = null;
        }
        this.havings = factory.and(exp.filterNotNull())
        return this
    }

    override fun having(exp: List<Predicate?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return having(*exp.toTypedArray())
        }
        this.havings = null;
        return this
    }

    override fun orderBy(vararg exp: Order?): CriteriaQuery<T>? {
        exp.filterNotNull()
            .forEach(orderBys::add)
        return this
    }

    override fun orderBy(exp: List<Order?>?): CriteriaQuery<T>? {
        if (exp != null) {
            return orderBy(*exp.toTypedArray())
        }
        return this
    }

    override fun distinct(d: Boolean): CriteriaQuery<T>? {
        this.isDistinct = d
        return this
    }

    override fun getOrderList(): List<Order?>? {
        return orderBys
    }

    override fun <X : Any?> from(cls: Class<X?>?): Root<X?>? {
        if (cls != null) {
            val root = factory.from(cls)
            roots.add(root)
            return root
        }
        TODO("Not yet implemented")
    }

    override fun <X : Any?> from(et: EntityType<X?>?): Root<X?>? {
        if (et != null) {
            val root = factory.from(et)
            roots.add(root)
            return root
        }
        TODO("Not yet implemented")
    }

    fun <X : Any?> from(query: CriteriaQuery<X>?): Root<X?>? {
        TODO("Not yet implemented")
    }

    override fun getRoots(): Set<Root<*>?>? {
        return roots
    }

    override fun getSelection(): Selection<T?>? {
        return selects
    }

    override fun getGroupList(): List<Expression<*>?>? {
        return groupBys.toList()
    }

    override fun getGroupRestriction(): Predicate? {
        return havings
    }

    override fun isDistinct(): Boolean {
        return this.isDistinct
    }

    override fun getResultType(): Class<T>? {
        return clazz
    }

    override fun <U : Any?> subquery(p0: Class<U?>?): Subquery<U?>? {
        TODO("Not yet implemented")
    }

    override fun <U : Any?> subquery(p0: EntityType<U?>?): Subquery<U?>? {
        TODO("Not yet implemented")
    }

    override fun getRestriction(): Predicate? {
        return wheres
    }

    override fun getParameters(): Set<ParameterExpression<*>?>? {
        TODO("Not yet implemented")
    }

    fun toSql(): String {
        return ""
    }
}